jxl/src/frame/modular/predict.rs

```diff
@@ -152,17 +155,247 @@ impl Predictor {
+        &mut self,
+        pos: (usize, usize),
+        xsize: usize,
+        data: &PredictionData,
+    ) -> (i64, i32) {
+        let (cur_row, prev_row) = if pos.1 & 1 != 0 {
+            (0, xsize + 2)
+        } else {
+            (xsize + 2, 0)
+        };
+        let pos_n = prev_row + pos.0;
+        let pos_ne = if pos.0 < xsize - 1 { pos_n + 1 } else { pos_n };
+        let pos_nw = if pos.0 > 0 { pos_n - 1 } else { pos_n };
+        let mut weights = [0u32; NUM_PREDICTORS];
+        for (i, weight) in weights.iter_mut().enumerate() {
+            *weight = error_weight(
+                self.pred_errors[i][pos_n]
+                    .wrapping_add(self.pred_errors[i][pos_ne])
+                    .wrapping_add(self.pred_errors[i][pos_nw]),
+                self.header.wp_header.w(i).unwrap(),
+            );
+        }
+        let n = add_bits(data.top);
+        let w = add_bits(data.left);
+        let ne = add_bits(data.topright);
+        let nw = add_bits(data.topleft);
+        let nn = add_bits(data.toptop);
+
+        let te_w = if pos.0 == 0 {
+            0
+        } else {
+            self.error[cur_row + pos.0 - 1]
+        };
+        let te_n = self.error[pos_n];
+        let te_nw = self.error[pos_nw];
+        let sum_wn = te_n + te_w;
+        let te_ne = self.error[pos_ne];
+
+        let p = if COMPUTE_PROPERTIES {
+            let abs_te_w = te_w.abs();
+            if te_n.abs() > abs_te_w {
+                te_n
+            } else if te_nw.abs() > abs_te_w {
+                te_nw
+            } else if te_ne.abs() > abs_te_w {
+                te_ne
+            } else {
+                te_w
+            }
+        } else {
+            0
+        } as i32;
+
+        self.prediction[0] = w + ne - n;
+        self.prediction[1] = n - (((sum_wn + te_ne) * self.header.wp_header.p1c as i64) >> 5);
+        self.prediction[2] = w - (((sum_wn + te_nw) * self.header.wp_header.p2c as i64) >> 5);
+        self.prediction[3] = n
+            - ((te_nw
+                .wrapping_mul(self.header.wp_header.p3ca as i64)
+                .wrapping_add(te_n.wrapping_mul(self.header.wp_header.p3cb as i64))
+                .wrapping_add(te_ne.wrapping_mul(self.header.wp_header.p3cc as i64))
+                .wrapping_add((nn - n).wrapping_mul(self.header.wp_header.p3cd as i64))
+                .wrapping_add((nw - w).wrapping_mul(self.header.wp_header.p3ce as i64)))
+                >> 5);
+
+        self.pred = weighted_average(&self.prediction, &mut weights);
+
+        if ((te_n ^ te_w) | (te_n ^ te_nw)) <= 0 {
+            let mx = w.max(ne.max(n));
+            let mn = w.min(ne.min(n));
+            self.pred = mn.max(mx.min(self.pred));
+        }
+        ((self.pred + PREDICTION_ROUND) >> PRED_EXTRA_BITS, p)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::headers::modular::{GroupHeader, WeightedHeader};
+
+    use super::{PredictionData, WeightedPredictorState};
+
+    struct SimpleRandom {
```

Comment:
Please add `rand` and say `rand_xoshiro` as a dev dependency instead.

Reply:
I generated the seed values for the libjxl golden data using this super simple method, and it doesn't feel useful to redo the work in libjxl _and_ add another dependency just for that.

Reply:
Ping